### DISCUSSION TITLE
Virtual Field (Resolved) Input Values

### DISCUSSION THREAD

**Discussion by ggpwnkthx** [keystonejs/keystone#9126] [Feature requests]
Is the spirit of the virtual field type designed strictly for calculating an output? I've run into an situation where having a calculated output and allow the user to submit data in the context of that field would be appropriate.

I have found a simple workaround is to use this change: https://github.com/keystonejs/keystone/blob/bd33f0785da2301ec6ff1e7e6ec3a135d6a3cd4c/packages/core/src/fields/types/virtual/index.ts#L76-L85
```typescript
        input: {
          create: {
            arg: graphql.arg({ type: graphql.JSON }), 
            resolve: () => undefined
          },
          update: { 
            arg: graphql.arg({ type: graphql.JSON }), 
            resolve: () => undefined
          },
        },
```
https://github.com/keystonejs/keystone/blob/bd33f0785da2301ec6ff1e7e6ec3a135d6a3cd4c/packages/core/src/fields/types/virtual/index.ts#L86-L89

And to change the existing `serialize` method defined here: https://github.com/keystonejs/keystone/blob/bd33f0785da2301ec6ff1e7e6ec3a135d6a3cd4c/packages/core/src/fields/types/virtual/views/index.tsx#L40-L51
```typescript
    serialize: value => {
      if (!value) return { [config.path]: null }
      try {
        return { [config.path]: value }
      } catch (e) {
        return { [config.path]: undefined }
      }
    },
```
https://github.com/keystonejs/keystone/blob/bd33f0785da2301ec6ff1e7e6ec3a135d6a3cd4c/packages/core/src/fields/types/virtual/views/index.tsx#L53-L54

This could definitely be fleshed out much better by defining actual `graphql.args` in the field schemas and handling those in the controller. I'm happy to tackle this, but I wanted to make sure this would make sense as a modification to the existing `virtual` field type or if a whole new `dynamic` field type would be better.

**Comment by ggpwnkthx** [comment]
I forgot to note that with the changes above, when using a custom `views` for the `virtual` field, the `inputData` in the `hooks` will be what ever value is passed into the `onChange` prop in the view.

**Comment by dcousens** [comment]
I think we could extend `virtual` to allow input, or something similar.
This is a very common use case and it has been on our internal backlog for quite some time.

**Reply by ggpwnkthx** [reply]
Thank you for the insight and confirmation of design intention. My team will work on this. Should I bother submitting a PR with the simple `graphql.JSON` implementation. We're using that as a default catch-all, but I can hold off on a PR until we have a more complete solution if that's preferred.

**Reply by ggpwnkthx** [reply]
#9139 submitted for review.
